### PR TITLE
lab1 - Implement Fibonacci and Negafibonacci

### LINKED ISSUE #312: Lab1 - Implement Fibonacci and Negafibonacci for davidbroughsmyth

### PR DESCRIPTION
**Pull request by davidbroughsmyth**
Fixes https://github.com/anz-bank/go-course/issues/312

Review of colleague's PR https://github.com/anz-bank/go-course/pull/321#pullrequestreview-247602940

### CODE REVIEW

**Review by prathish-ks — APPROVED** [review]
Looks good to me. In fact, it was helpful for me to review this with the number of test cases covered.

File: 01_fib/davidbroughsmyth/main.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+// Range of n from -92 .. 0 .. 92
+func fib(n int) {
+
+	if n == 0 {
+		fmt.Fprintln(out, 0)
+		return
+	}
+
+	if n > 92 || n < -92 {
+		fmt.Fprintf(out, "Range for fib(n) should be between -92 to 92\n")
```
Comment (anzboi) [inline 01_fib/davidbroughsmyth/main.go:21]: As a mathematician I would point out that technically `-92 to 92` is the domain, or the valid range of `n`, not `fib(n)` 😉

**Comment by anzboi** [comment]
I'm going to assume this has been reviewed by colleague (@sriyapaka)

File: 01_fib/davidbroughsmyth/main.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		return
+	}
+
+	var n2, n1, num int = 0, 1, n
+	if n < 0 {
+		num *= -1
+	}
+
+	for i := 0; i < num; i++ {
+		if n < 0 && i%2 != 0 {
```
Comment (anzboi) [inline 01_fib/davidbroughsmyth/main.go:31]: Very elegant way of keeping track of whether you should be printing negative numbers 👍

File: 01_fib/davidbroughsmyth/main_test.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	}
+}
+
+func TestFibNums(t *testing.T) {
+
+	tests := map[string]struct {
+		input int
+		want  string
+	}{
+		"fib(-1)":  {input: -1, want: strconv.Quote("1\n")},
```
Comment (anzboi) [inline 01_fib/davidbroughsmyth/main_test.go:30]: Great test coverage

**Review by anzboi — APPROVED** [review]
LGTM 👍 
I have edited your PR message to format it a little nicer. I would also recommend separating implementation and test in the PR comment before you merge.

File: 01_fib/davidbroughsmyth/main.go (hunk 1 above)
Comment (davidbroughsmyth) [reply 01_fib/davidbroughsmyth/main.go:21]: Good point, I was thinking the range of n but it does read for function fib(n).

**Pull request merged**